### DISCUSSION TITLE
How can I get mail working?

### DISCUSSION THREAD

**Discussion by einlichtvogel** [pocketbase/pocketbase#293] [Q&A]
Hey folks,

I just don't know what I am doing wrong, sending verification mail without SMPT on does not work, and sending emails with SMPT on and doublechecked SMPT credentials does also not work. 
Did anyone had the same Issue and could help me?

Thanks!

**Comment by ganigeorgiev** [comment]
Hi, In order to be able to send emails without SMTP, your system must have the `sendmail` process enabled and started (for windows there are applications that could emulate the same `sendmail` api). But even if `sendmail` is available it may still fail because the receiving server may reject it (that is often the case when sending emails to gmail accounts). On linux you could inspect the `sendmail` logs by running something like `cat /var/spool/mail/youruser`.

For the SMTP - to rule out an issue with the TLS encryption, please test again with both available options - _"Optional (StartTLS)"_ and _"Always"_.

For testing purposes I often use and recommend the free version of https://mailtrap.io/ (they also have a very convenient mail preview for both HTML and Text).

**Reply by ganigeorgiev** [reply]
You may also try starting the executable with the `--debug` flag and check if a more detailed error log is printed in the console, eg.:
```sh
./pocketbase serve --debug
```

**Reply by einlichtvogel** [reply]
I testet it with mailtrap and that works fine for me. However sending a mail over my own SMTP Server does not give any error in debug mode, but a eMail does not arrive at its destination.

**Reply by einlichtvogel** [reply]
I also tested it with the same settings as my mail program has, where I can send emails without any problems.
In pocketbase however it does not work. But since it works with mailtrap, the issue has to be on my side.
I'll look after it, thanks!

**Reply by ganigeorgiev** [reply]
Hm, that's strange. We are outputting any errors related to the mail connectivity when in `debug` mode.

Is it a personal SMTP server or you are using public service (eg. SendGrid, Mailgun, SES, etc.)? If possible could you try inspecting the mail server logs to see if a message was received.

Another thing that may prevent you to receive your email is that PocketBase has a very basic "spam/flood" protection that will skip sending the same email multiple times within a **2 minutes interval** (unless the user state hasn't changed before that) - here is an example where this is defined for the user verfication form - https://github.com/pocketbase/pocketbase/blob/master/forms/user_verification_request.go#L40 (also keep in mind that if your user is already verified then the verification endpoint will do nothing).

**Reply by einlichtvogel** [reply]
It's my friends WebHost but he's on vacation at the moment and I don't have any access to it.
Yeah I already realized that, it prints out in the debug console that the verification mail has already been send, but if you change the user like you said, or the mail settings, it works, and no error is there. When he's back I'll look after it with him.

**Reply by ganigeorgiev** [reply]
@blockiyt Just to let you know that with [v0.5.0 release](https://github.com/pocketbase/pocketbase/releases/tag/v0.5.0) there is now support for sending test emails from the Admin UI (the mailer errors are also added to the error message response as raw output for easier debugging). 

I've tested the current mailing implementation with 3 different providers (shared hosting mail server, mailgun, and mailtrap all with TLS on 587) and all works fine, but if you are still not able to connect to your SMTP server, feel free to let me know and I'll try to investigate it further.

**Reply by blujedis** [reply]
A little late to the part on this one but for posterity sake I'll leave this in the event someone stumbles upon it. Here are a few things to check.

1) Make sure wherever you're hosting your server that the provider doesn't have SMTP blocked. Linode for sure does this you must request the restriction be lifted. They and others started doing this a few years back for obvious reasons.

2) Ensure all DNS records are correct. Often DMARC issues can often arise and your mail could be simply rejected by your domain. If using Mailgun or similar this will be obvious in a quick look in the logs.

3) Be sure to have a valid EHLO set. Some servers will outright reject if this is not set correctly. 

4) Not a fan of StartTLS set to always. Read the spec you'll know why I said that. (preference)

5) Obviously username and password. While I sorted it quickly I had entered my account username rather than the SMTP relay account at my provider. 

Hope that saves someone a min or two!! :)
